File: amw/src/main/java/model/utils/LiteralBuilder.java
package model.utils;

import jason.asSyntax.*;

import java.util.*;

public class LiteralBuilder {

	private String name;
	private ListTerm values;
	private ListTerm queue;

	public LiteralBuilder( ) { }

	public LiteralBuilder( String s ) {
		this.name = s;
	}

	public LiteralBuilder setName( String s ) {
		this.name = s;
		return this;
	}

	public LiteralBuilder addValues( String... l ) {
		if ( values == null ) values = new ListTermImpl( );                         // first add
		for ( String s : l ) values.add( new Atom( s ) );                           // add all string
		return this;
	}

	public LiteralBuilder addValues( Term... terms ) {
		if ( values == null ) values = new ListTermImpl( );
		values.addAll( Arrays.asList( terms ) );
		return this;
	}

	public LiteralBuilder addQueue( String... l ) {
		if ( queue == null ) queue = new ListTermImpl( );
		for ( String s : l ) queue.add( new Atom( s ) );
		return this;
	}

	public LiteralBuilder addQueue( Term... terms ) {
		if ( queue == null ) queue = new ListTermImpl( );
		queue.addAll( Arrays.asList( terms ) );
		return this;
	}

	public Literal build( ) throws IllegalStateException {
		Literal literal = new LiteralImpl( new Atom( name ) );

		if ( values != null ) literal.addTerms( values.toArray( new Term[] {} ) );
		if ( queue != null ) literal.addAnnots( queue.toArray( new Term[] {} ) );

		return literal;
	}

	public static Literal buildLiteral( String name, String value ) {
		return buildLiteral( name, value, new Literal[] { } );
	}

	public static Literal buildLiteral( String name, String value, Literal... list ) {
		Literal literal = new LiteralImpl( new Atom( name ) );
		literal.addTerm( new Atom( value ) );

		literal.addAnnots( list );

		return literal;
	}

	public static Literal buildLiteral( String name, SimpleStructure... structures ) {
		return buildLiteral( name, structures, new Literal[] {} );
	}

	public static Literal buildLiteral( String name, SimpleStructure[] structures, Literal... list ) {
		Literal literal = new LiteralImpl( new Atom( name ) );

		Arrays.asList( structures ).forEach( entry -> {
			Structure struct = new Structure( entry.getName( ) );
			struct.addTerm( entry.getValue( ) );
			literal.addTerm( struct );
		} );

		literal.addAnnots( list );

		return literal;
	}

	public static class SimpleStructure {

		private String name;
		private Term value;

		public SimpleStructure( String name, Atom value ) {
			this.name = name;
			this.value = value;
		}

		public SimpleStructure( String name, String value ) {
			this.name = name;
			this.value = new Atom( value );
		}

		public SimpleStructure( String name, Integer value ) {
			this.name = name;
			this.value = new NumberTermImpl( value );
		}

		public String getName ( ) {
			return name;
		}

		public Term getValue ( ) {
			return value;
		}
	}
}

File: src/main/kotlin/common/translation/LiteralBuilder.kt
package common.translation

import jason.asSyntax.*

/**
 * An utility class created to easily build literals
 *
 * @author Paolo Baldini
 */
class LiteralBuilder(name: String) {
	companion object {
		fun pairTerm(name: String, value: String) = Structure(name).apply { addTerm(Atom(value)) }
		fun pairTerm(name: String, value: Double) = Structure(name).apply { addTerm(NumberTermImpl(value)) }
	}

	private val name = Atom(name)
	private val values: ListTerm by lazy { ListTermImpl() }
	private val queue: ListTerm by lazy { ListTermImpl() }

	/**
	 * Set values of the literal. A value is considered as:
	 *      name(VALUE, VALUE, ...)
	 */
	fun setValues(vararg l: String) = apply { values.addAll(l.map { Atom(it) }) }       // add all string

	/**
	 * Set values of the literal. A value is considered as:
	 *      name(VALUE, VALUE, ...)
	 */
	fun setValues(vararg l: Term) = apply { values.addAll(l) }                          // add all string

	/**
	 * Set queue of the literal. A queue-element is considered as:
	 *      name[QUEUE-ELEMENT, QUEUE-ELEMENT, ...]
	 */
	fun setQueue(vararg l:  String) = apply { queue.addAll(l.map { Atom(it) }) }        // add all string

	/**
	 * Set queue of the literal. A queue-element is considered as:
	 *      name[QUEUE-ELEMENT, QUEUE-ELEMENT, ...]
	 */
	fun setQueue(vararg l: Term) = apply { queue.addAll(l) }                            // add all string

	/**
	 * Assemble the literal
	 */
	fun build(): Literal {
		val literal = LiteralImpl(name)

		if (values.isNotEmpty()) literal.addTerms(*values.toTypedArray())
		if (queue.isNotEmpty()) literal.addAnnots(*queue.toTypedArray())

		return literal
	}
}